File: plugins/harness/skills/harness-install/assets/maven/harness-maven-plugin/src/main/java/ai/harness/maven/rules/DocContentRule.java
package ai.harness.maven.rules;

import ai.harness.maven.Finding;

import tools.jackson.databind.JsonNode;
import org.apache.maven.plugin.MojoExecutionException;
import java.nio.file.Path;
import java.util.ArrayList;
import java.util.Collection;
import java.util.List;
import java.util.stream.StreamSupport;

/**
 * Rule that requires specific content in documentation files.
 */
public enum DocContentRule implements HarnessCheckRule {
    INSTANCE;

    @Override
    public String category() {
        return "docContent";
    }
    private static final String CATEGORY = "docContent";

    @Override
    public boolean applies(JsonNode manifest) {
        return HarnessCheckHelper.applies(manifest, CATEGORY);
    }

    @Override
    public Collection<Finding> validate(Path root, JsonNode manifest) throws MojoExecutionException {
        final JsonNode catNode = manifest.get(CATEGORY);
        final JsonNode checksNode = catNode.get("parameters").get("checks");
        final String severity = HarnessCheckHelper.getSeverity(manifest, CATEGORY);
        return StreamSupport.stream(checksNode.spliterator(), false)
                .flatMap(check -> validateCheck(root, check, severity).stream())
                .toList();
    }

    private List<Finding> validateCheck(Path root, JsonNode check, String severity) {
        final String failureMessage = check.get("failureMessage").asText();
        final String combined = HarnessCheckHelper.extractPaths(check.get("files")).stream()
                .map(f -> {
                    try {
                        return HarnessCheckHelper.readFile(root, root.resolve(f));
                    } catch (MojoExecutionException e) {
                        return "";
                    }
                })
                .reduce((a, b) -> a + "\n" + b)
                .orElse("");
        return conditionMatches(check, combined)
                ? List.of()
                : List.of(new Finding(severity, CATEGORY, failureMessage));
    }

    private boolean conditionMatches(JsonNode check, String combined) {
        final JsonNode condition = check.has("condition") ? check.get("condition") : check.get("when");
        return condition == null
                ? false
                : evaluateCondition(condition, combined);
    }

    private boolean evaluateCondition(JsonNode condition, String combined) {
        if (condition == null || condition.isNull()) {
            return false;
        }
        if (condition.isTextual()) {
            return combined.contains(condition.asText());
        }
        if (condition.isArray()) {
            return StreamSupport.stream(condition.spliterator(), false)
                    .allMatch(item -> evaluateCondition(item, combined));
        }
        if (!condition.isObject()) {
            return false;
        }
        final boolean hasAll = condition.has("allOf");
        final boolean hasAny = condition.has("anyOf");
        final boolean hasContains = condition.has("contains");
        final boolean hasNot = condition.has("not");
        if (!(hasAll || hasAny || hasContains || hasNot)) {
            return false;
        }
        final JsonNode allOfValue = condition.get("allOf");
        final JsonNode anyOfValue = condition.get("anyOf");
        final List<JsonNode> allOf = conditionArray(allOfValue);
        final List<JsonNode> anyOf = conditionArray(anyOfValue);
        final List<String> contains = stringArray(condition.get("contains"));
        final boolean andMatches = allOf.isEmpty() || allOf.stream().allMatch(item -> evaluateCondition(item, combined));
        final boolean orMatches = !hasAny || !anyOf.isEmpty() && anyOf.stream().anyMatch(item -> evaluateCondition(item, combined));
        final boolean containsMatches = contains.stream().allMatch(combined::contains);
        final JsonNode notCondition = condition.get("not");
        final boolean notMatches = !hasNot || !evaluateCondition(notCondition, combined);
        return andMatches && orMatches && containsMatches && notMatches;
    }

    private List<JsonNode> conditionArray(JsonNode node) {
        if (node == null || node.isNull()) {
            return List.of();
        }
        if (!node.isArray()) {
            return List.of(node);
        }
        final List<JsonNode> items = new ArrayList<>();
        node.forEach(items::add);
        return items;
    }

    private List<String> stringArray(JsonNode node) {
        if (node == null || node.isNull()) {
            return List.of();
        }
        if (node.isTextual()) {
            return List.of(node.asText());
        }
        if (!node.isArray()) {
            return List.of();
        }
        final List<String> items = new ArrayList<>();
        node.forEach(item -> {
            if (item.isTextual()) {
                items.add(item.asText());
            }
        });
        return items;
    }
}

File: plugins/harness/skills/harness-install/assets/gradle/buildSrc/src/main/kotlin/ai/harness/gradle/rules/DocContentRule.kt
package ai.harness.gradle.rules

import ai.harness.gradle.HarnessPsiResults.Finding
import ai.harness.gradle.HarnessCheck
import kotlinx.serialization.json.JsonElement
import kotlinx.serialization.json.JsonObject
import kotlinx.serialization.json.contentOrNull
import kotlinx.serialization.json.jsonArray
import kotlinx.serialization.json.jsonObject
import kotlinx.serialization.json.jsonPrimitive
import java.nio.file.Path

/**
 * Rule that requires documentation files to contain specified content.
 */
object DocContentRule : HarnessCheckRule() {
    override val category: String = "docContent"
    override fun applies(manifest: JsonObject): Boolean =
        manifest[category]?.jsonObject?.get("enabled")?.jsonPrimitive?.contentOrNull?.toBoolean() ?: true

    override fun validate(
        manifest: JsonObject,
        root: Path,
    ): Collection<Finding> {
        val catObj = manifest[category]?.jsonObject
        val parametersObj = catObj?.get("parameters")?.jsonObject
        val checks = parametersObj?.get("checks")?.jsonArray
        return when {
            catObj == null || parametersObj == null || checks == null -> {
                emptyList()
            }

            else -> {
                checks
                    .mapNotNull { checkElem -> runCatching { checkElem.jsonObject }.getOrNull() }
                    .filter { checkObj ->
                        val files = HarnessCheck.stringArrayFrom(checkObj, "files")
                        val content =
                            files
                                .map { filePath ->
                                    HarnessCheck.readSafe(root, filePath)
                                }.joinToString("\n")
                        !conditionMatches(checkObj, content)
                    }.map { checkObj ->
                        Finding(
                            HarnessCheck.severityOf(manifest, category),
                            category,
                            HarnessCheck.stringFrom(checkObj, "failureMessage"),
                        )
                    }
            }
        }
    }

    private fun conditionMatches(
        checkObj: JsonObject,
        content: String,
    ): Boolean {
        val condition = checkObj["condition"] ?: checkObj["when"]
        return when {
            condition != null -> evaluateCondition(condition, content)
            else -> false
        }
    }

    private fun evaluateCondition(
        condition: JsonElement,
        content: String,
    ): Boolean {
        val asString = runCatching { condition.jsonPrimitive.contentOrNull }.getOrNull()
        if (asString != null) {
            return content.contains(asString)
        }
        val asArray = runCatching { condition.jsonArray }.getOrNull()
        if (asArray != null) {
            return asArray.all { item -> evaluateCondition(item, content) }
        }
        val asObject = runCatching { condition.jsonObject }.getOrNull() ?: return false
        val hasAll = "allOf" in asObject
        val hasAny = "anyOf" in asObject
        val hasContains = "contains" in asObject
        val hasNot = "not" in asObject
        if (!hasAll && !hasAny && !hasContains && !hasNot) {
            return false
        }
        val allOfValue = asObject["allOf"]
        val anyOfValue = asObject["anyOf"]
        val allOf = conditionArray(allOfValue)
        val anyOf = conditionArray(anyOfValue)
        val contains = stringArray(asObject["contains"])
        val andMatches = allOf.isEmpty() || allOf.all { item -> evaluateCondition(item, content) }
        val orMatches = !hasAny || (anyOf.isNotEmpty() && anyOf.any { item -> evaluateCondition(item, content) })
        val containsMatches = contains.all { item -> content.contains(item) }
        val notCondition = asObject["not"]
        val notMatches = !hasNot || notCondition?.let { item -> !evaluateCondition(item, content) } == true
        return andMatches && orMatches && containsMatches && notMatches
    }

    private fun conditionArray(value: JsonElement?): List<JsonElement> {
        val asArray = value?.let { item -> runCatching { item.jsonArray }.getOrNull() }
        return when {
            asArray != null -> asArray.toList()
            value != null -> listOf(value)
            else -> emptyList()
        }
    }

    private fun stringArray(value: JsonElement?): List<String> {
        val primitive = value?.let { item -> runCatching { item.jsonPrimitive.contentOrNull }.getOrNull() }
        val array = value?.let { item -> runCatching { item.jsonArray }.getOrNull() }
        return when {
            primitive != null -> listOf(primitive)
            array != null -> array.mapNotNull { item -> runCatching { item.jsonPrimitive.contentOrNull }.getOrNull() }
            else -> emptyList()
        }
    }

    override fun renderPsiFindings(
        findings: List<ai.harness.gradle.PsiFinding>,
        manifest: JsonObject,
    ): Collection<Finding> = emptyList()

    override fun findPsiFindings(
        file: Path,
        root: Path,
        psiFactory: org.jetbrains.kotlin.psi.KtPsiFactory?,
    ): List<ai.harness.gradle.PsiFinding> = emptyList()
}
